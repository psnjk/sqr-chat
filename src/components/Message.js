import React from "react";
import PropTypes from "prop-types";
import { formatRelative } from "date-fns";

import cn from "./Message.module.scss";

const formatDate = (date) => {
  let formattedDate = "";
  if (date) {
    // Convert the date in words relative to the current date
    formattedDate = formatRelative(date, new Date());
    // Uppercase the first letter
    formattedDate =
      formattedDate.charAt(0).toUpperCase() + formattedDate.slice(1);
  }
  return formattedDate;
};

const Message = ({
  createdAt = null,
  text = "",
  displayName = "",
  photoURL = "",
}) => {
  if (!text) return null;
  return (
    <div className={cn.root} data-testid="message">
      {photoURL ? (
        <img
          src={photoURL}
          alt="Avatar"
          className={cn.img}
          width={45}
          height={45}
        />
      ) : (
        <div className={cn.circle} />
      )}
      <div>
        <div className={cn.message}>
          {displayName ? <p className={cn.name}>{displayName}</p> : null}
          {createdAt?.seconds ? (
            <span className={cn.date}>
              {formatDate(new Date(createdAt.seconds * 1000))}
            </span>
          ) : null}
        </div>
        <p className={cn.text} data-testid="message-text">
          {text}
        </p>
      </div>
    </div>
  );
};

Message.propTypes = {
  text: PropTypes.string,
  createdAt: PropTypes.shape({
    seconds: PropTypes.number,
  }),
  displayName: PropTypes.string,
  photoURL: PropTypes.string,
};

export default Message;
