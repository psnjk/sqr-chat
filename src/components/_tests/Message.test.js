import React from "react";
import { act } from "react-dom/test-utils";
import renderer from "react-test-renderer";
import { render, cleanup } from "@testing-library/react";

import Message from "../Message";

afterEach(cleanup);

it("it renders without crashing", () => {
  render(<Message test={null}></Message>);
});

// it("it renders null", () => {
//   const { getByTestId } = render(
//     <Message
//       createdAt={{
//         nanoseconds: 874000000,
//         seconds: 1652009615,
//       }}
//       text={null}
//       displayName="My Name"
//       photoURL={undefined}
//     ></Message>
//   );
//   expect(getByTestId("message")).toBeNull();
//   // expect
// });

it("it renders correctly", () => {
  const { getByTestId } = render(
    <Message
      createdAt={{
        nanoseconds: 874000000,
        seconds: 1652009615,
      }}
      text="message text"
      displayName="My Name"
      photoURL={undefined}
    ></Message>
  );

  expect(getByTestId("message-text")).toHaveTextContent("message text");
});

it("matches snapshot", () => {
  const tree = renderer
    .create(
      <Message
        createdAt={{
          nanoseconds: 874000000,
          seconds: 1652009615,
        }}
        text="message text"
        displayName="My Name"
        photoURL={undefined}
      />
    )
    .toJSON();
  expect(tree).toMatchSnapshot();
});
