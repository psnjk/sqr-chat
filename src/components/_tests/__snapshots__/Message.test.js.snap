// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`matches snapshot 1`] = `
<div
  className="root"
  data-testid="message"
>
  <div
    className="circle"
  />
  <div>
    <div
      className="message"
    >
      <p
        className="name"
      >
        My Name
      </p>
      <span
        className="date"
      >
        Yesterday at 2:33 PM
      </span>
    </div>
    <p
      className="text"
      data-testid="message-text"
    >
      message text
    </p>
  </div>
</div>
`;
